node/src/testing/multi_stage_test_reactor/test_chain.rs

```diff
@@ -0,0 +1,365 @@
+            network,
+            gossip: gossiper::Config::new_with_small_timeouts(),
+            ..Default::default()
+        };
+
+        // ...and the secret key for our validator.
+        validator_config.consensus.secret_key_path = External::from_value(secret_key);
+
+        // Set a trust hash if one has been provided.
+        validator_config.node.trusted_hash = trusted_hash;
+
+        // Additionally set up storage in a temporary directory.
+        let (storage_config, temp_dir) = storage::Config::default_for_tests();
+        validator_config.consensus.unit_hashes_folder = temp_dir.path().to_path_buf();
+        self.storages.push(temp_dir);
+        validator_config.storage = storage_config;
+
+        // Bundle our config with a chainspec for creating a multi-stage reactor
+        let config = InitializerReactorConfigWithChainspec {
+            config: WithDir::new(&*CONFIG_DIR, validator_config),
+            chainspec: self.chainspec.clone(),
+        };
+
+        // Add the node (a multi-stage reactor) with the specified config to the network
+        self.network
+            .add_node_with_config(config, rng)
+            .await
+            .expect("could not add node to reactor")
+            .0
+    }
+}
+
+/// Get the set of era IDs from a runner.
+fn era_ids(runner: &Runner<ConditionCheckReactor<MultiStageTestReactor>>) -> HashSet<EraId> {
+    if let Some(consensus) = runner.reactor().inner().consensus() {
+        consensus.active_eras().keys().cloned().collect()
+    } else {
+        HashSet::new()
+    }
+}
+
+/// Given an era number, return a predicate to check if all of the nodes are in that specified era.
+fn is_in_era(era_num: usize) -> impl Fn(&Nodes<MultiStageTestReactor>) -> bool {
+    move |nodes: &Nodes<MultiStageTestReactor>| {
+        // check ee's era_validators against consensus' validators
+        let first_node = nodes.values().next().expect("need at least one node");
+
+        // Get a list of eras from the first node.
+        let expected_eras = era_ids(&first_node);
+
+        // Return if not in expected era yet.
+        if expected_eras.len() <= era_num {
+            return false;
+        }
+
+        // Ensure eras are all the same for all other nodes.
+        nodes
+            .values()
+            .map(era_ids)
+            .all(|eras| eras == expected_eras)
+    }
+}
+
+#[tokio::test]
+async fn run_validator_network() {
+    testing::init_logging();
+
+    let mut rng = crate::new_rng();
+
+    // Instantiate a new chain with a fixed size.
+    const NETWORK_SIZE: usize = 5;
+    let mut chain = TestChain::new(NETWORK_SIZE, &mut rng).await;
+
+    // Wait for all nodes to agree on one era.
+    for era_num in 1..=2 {
+        info!("Waiting for Era {} to end", era_num);
+        chain
+            .network
+            .settle_on(&mut rng, is_in_era(era_num), Duration::from_secs(600))
+            .await;
+    }
+}
+
+#[tokio::test]
+async fn run_equivocator_network() {
```

Comment:
I'm not seeing how a node equivocates here - could you explain please?

Reply:
This creates two nodes with the same private key, resulting in them double signing units and getting slashed and ultimately removed from the auction contract.

I have added a comment in this commit: https://github.com/CasperLabs/casper-node/pull/875/commits/46d2c808cd0716dc0ffa8cf96772ceaca41f9477

Reply:
Thanks - makes sense now!